src/unix/spawn-helper.cc

```diff
@@ -0,0 +1,63 @@
+#include <fcntl.h>
+#include <signal.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/resource.h>
+#include <termios.h>
+#include <unistd.h>
+
+#include "comms.h"
+
+int main (int argc, char** argv) {
+  sigset_t empty_set;
+  sigemptyset(&empty_set);
+  pthread_sigmask(SIG_SETMASK, &empty_set, nullptr);
+
+  struct rlimit rlim_ofile;
+  getrlimit(RLIMIT_NOFILE, &rlim_ofile);
+  for (rlim_t fd = STDERR_FILENO + 1; fd < rlim_ofile.rlim_cur; fd++) {
+    if (fd != COMM_PIPE_FD) {
+      close(fd);
+    }
+  }
+
+  setsid();
+
+#if defined(TIOCSCTTY)
+  // glibc does this
+  if (ioctl(STDIN_FILENO, TIOCSCTTY, NULL) == -1) {
+    _exit(1);
+  }
+#else
+  char *slave_path = ttyname(STDIN_FILENO);
+  // open implicit attaches a process to a terminal device if:
+  // - process has no controlling terminal yet
+  // - O_NOCTTY is not set
+  close(open(slave_path, O_RDWR));
+#endif
+
+  char *cwd = argv[0];
+  int uid = std::stoi(argv[1]);
+  int gid = std::stoi(argv[2]);
+  char *file = argv[3];
+  argv = &argv[3];
+
+  fcntl(COMM_PIPE_FD, F_SETFD, FD_CLOEXEC);
+
+  if (strlen(cwd) && chdir(cwd) == -1) {
+    perror("chdir(2) failed.");
+    _exit(1);
```

Comment:
Maybe give the faulty exits different exit codes, so we could separate them in parent and provide more explaining error messages.

Should we stick with perror here? This will end up on the pty master side (you will see it in terminus frontend), but cannot be intercepted programmatically in the parent script yet. I'd say any setup errors should be routed to parent, not to the pty.